### ISSUE TITLE
Feature Request: Mask groups using other geometry (world-space alpha maps).

### ISSUE THREAD

**Issue by adevart** [mrdoob/three.js#17771]
##### Description of the problem

There is a very useful object property colorWrite described here:

https://stackoverflow.com/questions/28869268/three-js-transparent-object-occlusion
http://jsfiddle.net/4vnsbdz6/1/
http://jsfiddle.net/4vnsbdz6/4/

However, it relies on manually setting the render order and excludes rendering of objects behind other objects.

What I would find useful is a rendering option that does this kind of occlusion but for object groups inside or perhaps just behind another geometry and using its alpha map to determine the visibility of the objects inside, with the option to set an influence value to do inverse.

For example, there would be a cube with an alpha map on it.
There would be a THREE.Group with a property that uses one or more objects as a mask with a mask influence such as the cube.
If the group is inside the cube, it would render the group objects according to the alpha map.
The group would have the list of mask objects and if any of the pixels in the group are behind an opaque pixel in the cube, they'd be rendered, otherwise not rendered.

This would behave like world-space alpha maps. It can probably be achieved using a custom shader but the shader would need to be applied per object. This would allow easily doing things like cutaways for architectural models:

https://cdn1.28dayslater.co.uk/xf/2017/07/1458657_8291f7d765e875b0ce71ac9087b1350b.jpg

For that, a full architecture geometry would be modelled and put in a group and then one or more masking geometries made with alpha maps. The architecture group would set the masking geometries as properties but with a negative influence so that only pixels outside the masking geometry are drawn fully and inside, according to the alpha map.

I guess it can be applied per material using a loop, it would just be easier to do per group.

The colorWrite examples above are almost usable for this but it's not possible to set an inverse on it, it doesn't use alpha maps so would have to model intricate masks but the most important aspect is being able to easily exclude or include geometry in the occlusion - being able to set a list of masks on a specific set of objects would be much easier than using render order, especially with multiple masks.

##### Three.js version

- [ ] Dev
- [x ] r109
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

##### Hardware Requirements (graphics card, VR Device, ...)

**Comment by adevart** [comment]
Being able to use custom geometry for clipping would also come close to achieving this effect (could flip the normals for inverse). Clipping planes work quite well for basic exclusions but are hard to position and can't be used for complex clipping. For using with object co-ordinates, it needs an update loop to reposition the planes.

**Comment by mrdoob** [comment]
TL;DR?

**Comment by adevart** [comment]
Short version is I'd like to be able to mask objects using other geometry either on the shader level using its textures or using the geometry itself like clipping planes (clipping planes can't do internal clipping).

Imagine a 3D building and having a cube at one of its corners. Everything inside the cube would behave like the colorWrite = false example and not render. Everything outside the cube and behind it would render as normal if this cube was set as a mask property. The output would be that you'd see inside the building and could animate this mask.

It's a 3D equivalent of 2D layer masking like in Photoshop. In Photoshop, you can mask an image and only show the bottom half. In 3D, it would be useful to do the same like take a 3D model and only show what's inside a selected 3D geometry, which would be its mask object.

In a game, a character can emerge from a portal.

This effect could also be produced using an alphaMap that can use a different co-ordinate system like world-space.

**Comment by adevart** [comment]
Here's an example of a portal effect:

https://i.postimg.cc/wBJNZ5tx/portal.jpg

The character would be a 3D mesh (duplicate characters, one at each portal), the white cubes would be the mask objects. None of the character mesh inside the cubes would render but everything else would like background and other objects in the masked space.

Clipping planes could work for this example but being able to use standard geometry works for more scenarios and easier to setup object-space transforms.
Render textures can give some flexibility with masking but there's a loss in quality with a render texture and they are hard to setup.
Setting colorWrite = false on the white cubes would also be an option but it requires everything else to have a different renderOrder, which complicates multiple masks.

**Comment by WestLangley** [comment]
>clipping planes can't do internal clipping

You _can_ clip the intersection of clipping planes. See https://threejs.org/examples/webgl_clipping_intersection.html.

>Clipping planes could work for this example but being able to use standard geometry works for more scenarios and easier to setup object-space transforms.

You can apply a transform to your clipping planes like so:

```js
var m = new THREE.Matrix4().makeRotationY( Math.PI / 2 );

var clipPlanes = [
	new THREE.Plane( new THREE.Vector3( 1, 0, 0 ), 0 ).applyMatrix4( m ),
	new THREE.Plane( new THREE.Vector3( 0, - 1, 0 ), 0 ).applyMatrix4( m ),
	new THREE.Plane( new THREE.Vector3( 0, 0, - 1 ), 0 ).applyMatrix4( m )
];
```

**Comment by adevart** [comment]
Ah, the intersection option looks like it would suffice, I wasn't aware of that. Thanks for posting that, I'll do some tests with it.

Is there a reason that clipping is done using THREE.Plane rather than say THREE.PlaneGeometry or any kind of geometry? Performance reasons maybe or does it make the calculations easier? Animated clipping in object-space needs the clipping plane transforms updated. Standard geometry can be parented to animated objects.

Maybe I can make a wrapper for it to create clipping planes from geometry and it would just create planes for each geometry face and use the intersection option so it behaves like a shape and just flip the normals for inverse. The class can store a reference to an object to use as its transform matrix.

How easy would it be to add feathering to the clipping edge? Just now it creates a hard edge. It would be good to set a soft falloff.

**Comment by EliasHasle** [comment]
> Maybe I can make a wrapper for it to create clipping planes from geometry and it would just create planes for each geometry face and use the intersection option so it behaves like a shape and just flip the normals for inverse.

It wouldn't work in general for non-convex geometries, right? Besides, I doubt three.js handles hundreds or thousands of clipping planes well. They will fill up the capacity for uniforms, and cause a lot of extra dot product operations per fragment.

I think masking based on geometry is a fascinating thought, though.

**Comment by adevart** [comment]
Would it be possible to have a function that exposed the color, screen space and world space co-ordinates of an object's rendered pixels to allow a program to manipulate the output?

For example, an object renders to screen and there would be a function per object like the following that runs for every pixel sample that is rendered to screen:

renderOutput(pixel) {
screenPosition = pixel.screenPosition;
worldPosition = pixel.worldPosition;
pixelColor = pixel.color;

// in here, the returned pixel color could be manipulated
// based on different attributes like worldPosition e.g
// if pixel.worldPosition inside masking geometry, pixelColor.alpha = 0;

return pixelColor;
}

This would be like a compositing shader. Ideally not webGL but maybe that would be too slow due to the number of samples. This would have to be executed in such a way that if pixels closest to the camera had their opacity changed, the rasterizer would render pixels behind.

With a function like this, any kind of masking would be possible. That function could have its own test for where the world-space co-ordinates were for a rendered pixel and adjust the opacity of it and it could soften the edges. Screen-space and world-space co-ordinates allows for view-dependent and independent effects.

Maybe just a custom implementation of the clipping fragment shader would work:

https://github.com/mrdoob/three.js/blob/dev/src/renderers/shaders/ShaderChunk/clipping_planes_fragment.glsl.js

Instead of looping over clipping planes, it would loop over the faces of some geometry and determine where the world-space position of the fragment was relative to each face and change its opacity depending on the face normal. For a soft edge, it can calculate the distance from the closest geometry face to do a gradual falloff.

**Comment by EliasHasle** [comment]
Considering @mrdoob's "TL;DR" comment above, and the rather speculative nature of your proposals, I suggest opening a discussion on the [forum](https://discourse.threejs.org) about this.

**Comment by Mugen87** [comment]
> Considering @mrdoob's "TL;DR" comment above, and the rather speculative nature of your proposals, I suggest opening a discussion on the forum about this.

👍
